src/hardware/tplink-hs/protocol.c

```diff
@@ -0,0 +1,491 @@
+	if (tplink_hs_get_node_value(resp, "sw_ver",
+				       &devc->dev_info.sw_ver) != SR_OK)
+		goto err;
+	if (tplink_hs_get_node_value(resp, "deviceId",
+				       &devc->dev_info.device_id) != SR_OK)
+		goto err;
+
+	g_free(resp);
+
+	sr_spew("Registered device: %s - %s - %s", devc->dev_info.model,
+						   devc->dev_info.sw_ver,
+						   devc->dev_info.device_id);
+
+	return SR_OK;
+
+err:
+	g_free(devc->dev_info.model);
+	g_free(devc->dev_info.sw_ver);
+	g_free(devc->dev_info.device_id);
+	g_free(resp);
+
+	return SR_ERR;
+}
+
+static void handle_poll_data(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc = sdi->priv;
+	struct sr_datafeed_packet packet;
+	struct sr_datafeed_analog analog;
+	struct sr_analog_encoding encoding;
+	struct sr_analog_meaning meaning;
+	struct sr_analog_spec spec;
+	int i;
+
+	sr_analog_init(&analog, &encoding, &meaning, &spec, 0);
+
+	packet.type = SR_DF_ANALOG;
+	packet.payload = &analog;
+	analog.num_samples = 1;
+
+	for (i = 0; devc->dev_info.channels[i].name; i++) {
+		analog.meaning->mq = devc->dev_info.channels[i].mq;
+		analog.meaning->unit = devc->dev_info.channels[i].unit;
+		analog.meaning->mqflags = SR_MQFLAG_DC;
+		analog.encoding->digits = 6;
+		analog.spec->spec_digits = 6;
+
+		if (devc->dev_info.channels[i].mq == SR_MQ_VOLTAGE) {
+			analog.meaning->channels =
+				g_slist_append(NULL, sdi->channels->data);
+			analog.data = &devc->voltage;
+		}
+		else if (devc->dev_info.channels[i].mq == SR_MQ_CURRENT) {
+			analog.meaning->channels =
+				g_slist_append(NULL, sdi->channels->next->data);
+			analog.data = &devc->current;
+		}
+
+		sr_session_send(sdi, &packet);
+	}
+	sr_sw_limits_update_samples_read(&devc->limits, 1);
+}
+
+static int recv_poll_data(struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc = sdi->priv;
+	char *response = g_malloc0(1024);
+	char *node_current_value;
+	char *node_voltage_value;
+	int len;
+
+	len = tplink_hs_tcp_read_data(devc, response, 1024);
+
+	if (len < 0)
+		goto err;
+
+	if (tplink_hs_get_node_value(response, "current",
+			       &node_current_value) != SR_OK)
+		goto err;
+	if (tplink_hs_get_node_value(response, "voltage",
+			       &node_voltage_value) != SR_OK)
+		goto err;
+
+	sr_spew("volatage: %s, current: %s", node_voltage_value,
+					     node_current_value);
+
+	devc->voltage = strtof(node_voltage_value, NULL);
+	devc->current = strtof(node_current_value, NULL);
+
+	sr_spew("volatage(f): %f, current(f): %f", devc->voltage,
+						    devc->current);
+
+	handle_poll_data(sdi);
+
+	g_free(response);
+	return SR_OK;
+
+err:
+	g_free(response);
+	return SR_ERR;
+
+}
+
+SR_PRIV int tplink_hs_receive_data(int fd, int revents, void *cb_data)
+{
+	struct sr_dev_inst *sdi;
+	struct dev_context *devc;
+	int64_t now, elapsed;
+
+	(void)fd;
+
+	if (!(sdi = cb_data))
+		return TRUE;
+
+	if (!(devc = sdi->priv))
+		return TRUE;
+
+	if (revents == G_IO_IN) {
+		sr_info("In callback G_IO_IN");
+		recv_poll_data(sdi);
+		tplink_hs_tcp_close(devc);
```

Comment:
I'm pretty sure closing the socket here and keeping the session source pollfd is illegal and only works because my socket happens to get the same fd next time I connect, I'd like feedback on this.

I'm doing this because unfortunately the hs110 requires a new tcp connection for each requests and if I stay connected, the G_IO_IN event will continuously trigger. I also want to keep my event source so this callback will get called on timeout and so eventually I can poll the device again.

Reply:
I switched to a combination of timer along with a temporary registration of the socket to get the reply. Now it seems legit to me.